platform/android/java/lib/src/org/godotengine/godot/input/Joystick.java

```diff
@@ -30,16 +30,23 @@
 
 package org.godotengine.godot.input;
 
-import android.view.InputDevice.MotionRange;
-
 import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * POJO class to represent a Joystick input device.
  */
 class Joystick {
 	int device_id;
 	String name;
-	ArrayList<MotionRange> axes;
-	ArrayList<MotionRange> hats;
+	List<Integer> axes = new ArrayList<Integer>();
+	protected boolean hasAxisHat = false;
+	/*
+	 * Keep track of values so we can prevent flooding the engine with useless events.
+	 */
+	protected final Map<Integer, Float> axesValues = new ConcurrentHashMap<Integer, Float>();
```

Comment:
Does this map need to be thread safe?

Reply:
Are the calls into GodotInputHandler by the Android OS or Fire OS (etc) guaranteed to be single threaded?

Reply:
Yes, the calls are dispatched on the Android UI/main thread and are thus guaranteed to be single-threaded.

Reply:
Ok. Will change.